design/kata-2-0-metrics.md

```diff
@@ -0,0 +1,1343 @@
+
+There are some limitations, or called missing parts, the lack of observability may become an obstacles to run Kata containers in large scale.
+
+In Kata 2.0, these parts will be took to provide more details about the system.
+
+- Shim process(Kata runtime)
+- Hypervisor stats
+- Agent process
+- Guest OS stats
+
+## Metrics architecture
+
+Kata 2.0 metrics strongly depend on [Prometheus](https://prometheus.io/), a graduated project from CNCF.
+
+The new design will introduce a new command(component) named `kata-magent`, the `m` of `magent` means metrics or management, this command is used to manage/monitor Kata runtime it self other than containers. It's the management interface with Kata runtime, and we can do something like these:
+
+- Get metrics
+- Get events
+- Extend hypervisor commands
+
+In this document we will cover metrics only. And until now it only supports metrics function.
+
+This is the architecture overview metrics in Kata Containers 2.0.
+
+![Kata Containers 2.0 metrics](arch-images/kata-2-metrics.png)
+
+
+And the sequence diagram is shown below:
+
+![Kata Containers 2.0 metrics ](arch-images/kata-metrics-sequence-diagram.png)
+
+
+### Kata magent
+
+`magent` is a management agent on one nodes, where many Kata containers are running. `magent`'s work include:
+
+- Aggregate sandbox metrics running on this node, and add `sandbox_id` label
+- As a prometheus target, all metrics from Kata shim on this node will be collected by prometheus indirectly. This can easy the targets count in Prometheus, and also need not to expose shim's metrics by `ip:port`
+
+Only one `magent` process are running on one node.
+
+`magent` is using a different communication channel other than that `conatinerd` communicating with Kata shim, and Kata shim listen on a new socket address for communicating with `magent`.
+
+The way `magent` get shim's metrics socket file(`magent_address`) like that `containerd` get shim address. The socket is an abstract socket and saved as file `abstract` with the same directory of `address` for `containerd`.
+
+**Note**: If there is no Prometheus server is configurated, i.e., there is no scrape operations, `magent` will do nothing initiative.
+
+### Kata runtime
+
+Runtime are mainly to:
+
+- Gather metrics about shim process
+- Gather metrics about hypervisor process
+- Gather metrics about running sandbox
+- Get metrics from Kata agent(through `ttrpc`)
+
+### Kata agent
+
+Agent is responsible for:
+
+- Gather agent process metrics
+- Gather guest OS metrics
+
+And in Kata 2.0, agent will add a new interface:
+
+```
+rpc GetMetrics(GetMetricsRequest) returns (Metrics);
+
+message GetMetricsRequest {}
+
+message Metrics {
+	string metrics = 1;
+}
+
+```
+
+The `metrics` field is Prometheus encoded content. This can avoid defining a fixed struct in protocol buffers.
+
+### Performance and overhead
+
+Metrics should not become the bottleneck of system, downgrade the performance, and run with minimal overhead.
```

Comment:
I agree but I think you could probably summarise this as a brief set of requirements:

> Requirements:
>
> - Metrics MUST be quick to collect (can we guarantee this? Possibly not?)
> - Metrics MUST be small.
> - Metrics MUST be generated only if there are subscribers to the Kata metrics service (in other words, zero overhead if nobody cares ;)
> -> Metrics MUST NOT be cached by the Kata metrics service (to guarantee reliable results and to remove any possibility of DoS by a rogue client trying to kill the metrics service).
> - ...


Then, you could put the details in the table into the commit along with details of what type of system you ran them on, version of Kata components, *etc*.